### ISSUE TITLE
Chop up the package - v4.0

### ISSUE THREAD

**Issue by fprochazka** [Kdyby/Doctrine#238]
- kdyby/doctrine-dql-repositories (better name?)
    - `EntityRepository` (possibly as a trait, to allow mixing more repository features to default one, without extensive rape of inheritance)
    - customized `QueryBuilder` (maybe create another repo for this one and make the repositories repo depend on it?)
- kdyby/doctrine-query-object
    - `QueryObject`
        - build-in events for select & filter callback array
        - public api for "getting" the actualy generated query for custom purposes (or passing it somwhere else)
    - `ResultSet`
        - simplify
    - `EmptyResultSet`
- kdyby/doctrine-readonly-collection
- kdyby/doctrine-lazy-collection
- kdyby/doctrine-magic-accessors
- kdyby/doctrine-nette-reflection
    - `RuntimeReflectionService`
- kdyby/doctrine-non-locking-unique-inserter
- kdyby/doctrine-dbal-batch-execute
    - `Kdyby\Doctrine\Helpers::executeBatch()`
    - `Kdyby\Doctrine\Helpers::loadFromFile()`
- drop
    - Persistence namespace
    - geo
    - dao
    - custom dbal exceptions (solved long time ago natively)
    - autodiscovery of repository<=>entity, make it a configuration section without magic (it should configure listeners that will override metadata settings appropriately)
    - DqlSelection & Dql namespace ... as I'm not using it and I don't know about anybody who is
- kdyby/doctrine - OrmExtension
    - will have "load" sections for all the packages, so just installing them via composer should be enough to enable them (or most of them), but should work without any of them

**Comment by enumag** [comment]
I'd probably not chop it all the way to single-class packages if I were you but I don't really mind either. :+1: 

kdyby/doctrine-dql-repositories (better name?) - possibly just kdyby/doctrine-dql? By the way this will need some thinking because the EntityRepository contains a fetch method for QueryObject and you want to drop the Persistence namespace which contains an interface for QueryObject used by fetch. Possibly kdyby/doctrine-query-object would depend on this package? (Certainly not the other way around.)

EntityRepository as a trait - I'd maybe use multiple traits for the different features.

I don't use any of the features you want to drop. Or at least I don't think I am. So :+1: for that. :-)

**Comment by enumag** [comment]
Oh and anything console-related should probably be separated as well.

**Comment by enumag** [comment]
Also I'd like to be able to remove kdyby/events from my projects completely. Any chance of kdyby/doctrine not depending on it?

**Comment by juniwalk** [comment]
```
kdyby/doctrine-readonly-collection
kdyby/doctrine-lazy-collection
```
Why not just `kdyby/doctrine-collections` and merge them into one package?

**Comment by fprochazka** [comment]
> the EntityRepository contains a fetch method for QueryObject

That should be removed. It doesn't really make any sense to have the fetch method on repository, since PHP doesn't know generics. The `Persistence` namespace won't be completely dropped, the QO related stuff will probably be refined and moved to the QO-related repository.

> I'd maybe use multiple traits for the different features.

Exactly :) 

> Oh and anything console-related should probably be separated as well - and the kdyby/console dependency should be dropped. What do you think?

Discuss that here please https://github.com/Kdyby/Doctrine/issues/190.

> Also I'd like to be able to remove kdyby/events from my projects completely. Any chance of kdyby/doctrine not depending on it?

Definitely new issue, I might be in for an optional switch between symfony-something and kdyby/events... But... new issue :)

---

@juniwalk well, people are almost iracionally obsessed with minimalism theese days, but chopping it up doesn't really hurt anything (except maintenance cost)... Also, if it has strictly defined name like `doctrine-readonly-collection` the discussion where a new class, for example `SomeMagicCollectionsHelper` should be added here, or in new package doesn't even happen, since it cannot be added to the existing package to "contaminate" it.

**Comment by juniwalk** [comment]
@fprochazka quilty as charged, I do love minimalism :D

**Comment by TomasVotruba** [comment]
:+1:

**Comment by foxycode** [comment]
@fprochazka Can I help with something?

**Comment by fprochazka** [comment]
@foxycode well, the tasks require a lot of "management work", so if you wanna help, I'd encourage you to start with the code cleanup, which means dropping theese (preferably in several pullrequests, not in one)

- geo
- dao
- custom dbal exceptions

**Comment by juniwalk** [comment]
@fprochazka I think ancient PRs and issues that are no longer relevant should be cleared too. Also after chopping up some of them may become irrelevant as they would be for code in different repository.

**Comment by foxycode** [comment]
@fprochazka Ok, I look into it.

**Comment by foxycode** [comment]
@fprochazka I have first commit prepared. I'm not sure how to make PR separate from #243, Do I need to create own branch or is there simpler way?

**Comment by fprochazka** [comment]
@foxycode yes, creating PR's from master is an anti-pattern. You should create new branch for each PR.

**Comment by fprochazka** [comment]
If somebody wants to help, here are some usefull tasks https://github.com/Kdyby/DoctrineDbalBatchImport/issues

**Comment by enumag** [comment]
TODO: optional dependencies

- Events: https://github.com/Kdyby/Doctrine/issues/239 (PR https://github.com/Kdyby/Doctrine/pull/266)
- Console: https://github.com/Kdyby/Doctrine/issues/190
- Annotations: https://github.com/Kdyby/Doctrine/issues/268

**Comment by Olicek** [comment]
Is this still in progress?

**Comment by fprochazka** [comment]
It's still planned, but I have very little time currently.

**Comment by enumag** [comment]
@fprochazka I think we should split doctrine/dbal integration into its own package as well. Sometimes I want to only use the DBAL without the ORM. What do you think?

**Comment by fprochazka** [comment]
@enumag I'm not against it. Let me think about that.

**Comment by fprochazka** [comment]
TBD: throw away QO and suggest https://github.com/librette/doctrine-queries ?

**Comment by fprochazka** [comment]
* More clear separation of repository magic would be nice #182

**Comment by fprochazka** [comment]
If we ignore the extra functionality, that will be extracted to separate packages (I consider pretty straightforward problem), I have an updated proposal:

## The Plan

* minimum requirements:
    * PHP 7.1 and newer
    * Nette 2.4 (3.0 will be solved as separate major release, once it is actually released)
    * Current stable Doctrine packages (once the 7.1 only versions are released, all packages will increase the requirement to those versions with new major release probably)
* `Kdyby/NetteDoctrineAdapter`
    * `ManagerRegistry`
    * `ConnectionFactory`
    * `ProxyAutoloader` (for when entity proxies are store in sessions)
    * DI
        * `ContainerAwareEventManager`
        * `ContainerAwareEntityListenerResolver`
    * Caching
        * `CacheCleaner`
    * Tracy
        * `BluescreenRenderer` - all exception rendering that adds panels to bluescreen
        * `UnitOfWorkSnapshotRenderer` - currently the panel depends on wiring into `EntityManager` but if a listener was created, that would create snapshot of UoW before every flush, this dependency could be removed
        * `SecondLevelCacheStatsPanel`
    * and other similar classes required for bridging the packages
    * console commands proxies - same as [DoctrineBundle](https://github.com/doctrine/DoctrineBundle/tree/6e96577cbbdbb5b6dcca2ff203d665976b51beb0/Command/Proxy)
* `Kdyby/TracyBarDoctrineDbalAdapter`
   * a Tracy panel, that will handle **only** SQL queries and highlighting in the Tracy Bar
   * + secret idea to solve showing executed DQL in panel
* `Kdyby/LoggerDoctrineDbalAdapter`
   * a `SQLLogger` that will handle **only** SQL queries piping into `Psr\Log\LoggerInterface`
* `Kdyby/Doctrine`
   * will contain only `DoctrineExtension` that will have the same configuration structure as Symfony (`doctrine: dbal: ...` and `doctrine: orm: ...` sections)
   * the convention with Symfony is to have `under_score` config notation but most Nette extensions have `camelCase` notation => the structure should be the same, but `camelCase`
   * require
       * `Kdyby/NetteDoctrineAdapter`
       * `Kdyby/TracyBarDoctrineDbalAdapter`
    *  everything else will be optional

## Open questions:

### queries higlighting

idea:

* `Kdyby/TracyBarDoctrineDbalAdapter`
   * this lib will define custom `QueryFormatter` interface with a very simple implementation, that will only format parameters into the query with no hilighting
* `Kdyby/DoctrineDbalQueryFormatter`
   * a `SQLLogger` that will handle **only** SQL queries piping into `Psr\Log\LoggerInterface`

or

* just have it in `TracyBarDoctrineDbalAdapter`, but abstract it away with interface anyway to allow changing the behaviour

I will probably do it the second way because it's way easier
